src/foam/core/XML.js

```diff
@@ -0,0 +1,466 @@
+    },
+
+    {
+      name: 'outputProperty',
+      code: foam.mmethod({
+        String: function(v, p) { this.outputPrimitive(v, p) },
+        Number: function(v, p) { this.outputPrimitive(v, p) },
+        Boolean: function(v, p) { this.outputPrimitive(v, p) },
+        Date: function(v, p) { this.outputPrimitive(v, p) },
+        Array: function(v, p) {
+          this.start("<" + this.propertyName(p) + " class='Array'>");
+          this.nl().indent();
+          this.output(p.toXML(v, this));
+          this.end('</' +  this.propertyName(p) + '>');
+        },
+        FObject: function(v, p) {
+          this.start("<" + this.propertyName(p) + ">");
+          this.nl().indent();
+          this.output(p.toXML(v, this));
+          this.end('</' +  this.propertyName(p) + '>');
+        },
+        AbstractEnum: function(v, p) {
+          this.start("<" + this.propertyName(p) + ">");
+          this.outputProperty_(v, v.cls_.getAxiomByName('ordinal'));
+          this.end('</' +  this.propertyName(p) + '>');
+        }
+      })
+    },
+
+    function outputPrimitive(v, p){
+      this.out('<').outputPropertyName(p).out('>');
+      this.output(p.toXML(v, this));
+      this.out('</').outputPropertyName(p).out('>');
+    },
+
+    function outputDate(o) {
+      if ( this.formatDatesAsNumbers ) {
+        this.out(o.valueOf());
+      } else {
+        this.out(JSON.stringify(o));
+      }
+    },
+
+    function outputFunction(o) {
+      if ( this.formatFunctionsAsStrings ) {
+        this.output(o.toString());
+      } else {
+        this.out(o.toString());
+      }
+    },
+
+    function outputObjectKeyValue_(key, value, first) {
+      if ( ! first ) this.out(',').nl().indent();
+      this.out(this.maybeEscapeKey(key), ':').output(value);
+    },
+
+    function outputObjectKeyValues_(o) {
+      var first = true;
+      for ( var key in o ) {
+        this.outputObjectKeyValue_(key, o[key], first);
+        first = false;
+      }
+    },
+
+    function outputSortedObjectKeyValues_(o) {
+      var key, keys = [];
+
+      for ( key in o ) keys.push(key);
+      keys.sort();
+
+      var first = true;
+      for ( var i = 0 ; i < keys.length; i++ ) {
+        key = keys[i];
+        this.outputObjectKeyValue_(key, o[key], first);
+        first = false;
+      }
+    },
+
+    {
+      name: 'output',
+      code: foam.mmethod({
+        Undefined: function(o) { this.out('null'); },
+        Null:      function(o) { this.out('null'); },
+        String:    function(o) { this.out(this.escape(o)); },
+        Number:    function(o) { this.out(o); },
+        Boolean:   function(o) { this.out(o); },
+        Date:      function(o) { this.outputDate(o); },
+        Function:  function(o) { this.outputFunction(o); },
+        AbstractEnum: function(o) { },
+        FObject:   function(o) {
+          if ( o.outputXML ) {
+            o.outputXML(this)
+            return;
+          }
+
+          var clsName = o.cls_.id;
```

Comment:
We should pass clsName through a local method like "escapeAttribute" or "outputAttributeValue" that ensures that any nested quotes are escaped.  Not that we expect quotes in class id's but better to be safe.